exporter/prometheusremotewriteexporter/exporter.go

```diff
@@ -216,46 +220,67 @@ func (prwe *prwExporter) export(ctx context.Context, requests []*prompb.WriteReq
-	if err != nil {
-		return consumererror.NewPermanent(err)
+		// Create the HTTP POST request to send to the endpoint
+		req, err := http.NewRequestWithContext(ctx, "POST", prwe.endpointURL.String(), bytes.NewReader(compressedData))
+		if err != nil {
+			return backoff.Permanent(consumererror.NewPermanent(err))
+		}
+
+		// Add necessary headers specified by:
+		// https://cortexmetrics.io/docs/apis/#remote-api
+		req.Header.Add("Content-Encoding", "snappy")
+		req.Header.Set("Content-Type", "application/x-protobuf")
+		req.Header.Set("X-Prometheus-Remote-Write-Version", "0.1.0")
+		req.Header.Set("User-Agent", prwe.userAgentHeader)
+
+		resp, err := prwe.client.Do(req)
+		if err != nil {
+			return err
+		}
+		defer resp.Body.Close()
+
+		// 2xx status code is considered a success
+		// 5xx errors are recoverable and the exporter should retry
+		// Reference for different behavior according to status code:
+		// https://github.com/prometheus/prometheus/pull/2552/files#diff-ae8db9d16d8057358e49d694522e7186
+		if resp.StatusCode >= 200 && resp.StatusCode < 300 {
+			return nil
+		}
+
+		body, err := io.ReadAll(io.LimitReader(resp.Body, 256))
+		rerr := fmt.Errorf("remote write returned HTTP status %v; err = %w: %s", resp.Status, err, body)
+		if resp.StatusCode >= 500 && resp.StatusCode < 600 {
+			return rerr
+		}
+		return backoff.Permanent(consumererror.NewPermanent(rerr))
 	}
 
-	// Add necessary headers specified by:
-	// https://cortexmetrics.io/docs/apis/#remote-api
-	req.Header.Add("Content-Encoding", "snappy")
-	req.Header.Set("Content-Type", "application/x-protobuf")
-	req.Header.Set("X-Prometheus-Remote-Write-Version", "0.1.0")
-	req.Header.Set("User-Agent", prwe.userAgentHeader)
+	// Use the BackOff instance to retry the func with exponential backoff.
+	err := backoff.Retry(retryFunc, &backoff.ExponentialBackOff{
+		InitialInterval:     prwe.retrySettings.InitialInterval,
+		RandomizationFactor: prwe.retrySettings.RandomizationFactor,
+		Multiplier:          prwe.retrySettings.Multiplier,
+		MaxInterval:         prwe.retrySettings.MaxInterval,
+		MaxElapsedTime:      prwe.retrySettings.MaxElapsedTime,
+		Stop:                backoff.Stop,
+		Clock:               backoff.SystemClock,
+	})
 
-	resp, err := prwe.client.Do(req)
 	if err != nil {
 		return consumererror.NewPermanent(err)
 	}
-	defer resp.Body.Close()
 
-	// 2xx status code is considered a success
-	// 5xx errors are recoverable and the exporter should retry
-	// Reference for different behavior according to status code:
-	// https://github.com/prometheus/prometheus/pull/2552/files#diff-ae8db9d16d8057358e49d694522e7186
-	if resp.StatusCode >= 200 && resp.StatusCode < 300 {
-		return nil
-	}
-	body, err := io.ReadAll(io.LimitReader(resp.Body, 256))
-	rerr := fmt.Errorf("remote write returned HTTP status %v; err = %w: %s", resp.Status, err, body)
-	if resp.StatusCode >= 500 && resp.StatusCode < 600 {
-		return rerr
-	}
-	return consumererror.NewPermanent(rerr)
+	return nil
```

Comment:
I don't think this is correct.  I think it still needs to return `err`:

```suggestion
	return err
```

From the [godoc](https://pkg.go.dev/github.com/cenkalti/backoff/v4#Retry):

> If o returns a *PermanentError, the operation is not retried, and the wrapped error is returned. 

So the `err` returned will be either a `consumererror.Permanent` wrapping a non-retryable error or it will be a retryable error that will not be retried as the `Backoff` instance has hit a timeout or other stopping condition, or it will be `nil` if the operation succeeded.  In any of those cases it can be returned correctly from this method.

Reply:
actually, it can be either `err` or `nil` because anything other than nil will always be permanent error just using same [error propagation form like before](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/23842/files#diff-2631d8a8855fdc08cf077c0ab99544147ca73eec572bc871d8fe717e826eda38L258).
 At this line it won't be a retryable error anymore, the only retryable error is from this check `resp.StatusCode >= 500 && resp.StatusCode < 600` and retries are done.
 
Even, if we want to send as it is from the backoff without wrapping in [permanent error](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/23842/files#diff-2631d8a8855fdc08cf077c0ab99544147ca73eec572bc871d8fe717e826eda38R280), like a retryable error that will not be retried as the Backoff instance has hit a timeout (in any case) it is already a [consumererror.Permanent](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/23842/files#diff-2631d8a8855fdc08cf077c0ab99544147ca73eec572bc871d8fe717e826eda38R210) in export.

Reply:
Returning `nil` is not correct in the case that a permanent error or timeout was encountered.  Yes, it is correct that there is no further action this component can take with such an error, but it may be important to pass that error back up the call stack so that a caller can appropriately react to the error.

Reply:
Agreed, committed suggestion